src/SAML2/WebSSO/Test/Lenses.hs

```diff
@@ -0,0 +1,35 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | WARNING: these optics make assumptions about the shape of the AuthnResponse that are not valid
+-- in general!  (they are useful for testing, though.)
+module SAML2.WebSSO.Test.Lenses where
+
+import Control.Lens
+import Data.List.NonEmpty as NL
+import SAML2.WebSSO
+import Test.QuickCheck.Instances ()
+
+
+_nlhead :: Lens' (NonEmpty a) a
+_nlhead f (a :| as) = (:| as) <$> f a
+
+assertionL :: Lens' AuthnResponse Assertion
+assertionL = rspPayload . _nlhead
+
+conditionsL :: Traversal' AuthnResponse Conditions
+conditionsL = assertionL . assConditions . _Just
+
+subjL :: Lens' AuthnResponse Subject
+subjL = assertionL . assContents . sasSubject
+
+scdataL :: Traversal' AuthnResponse SubjectConfirmationData
+scdataL = subjL . subjectConfirmations . ix 0 . scData . _Just
+
+statementL :: Lens' AuthnResponse Statement
+statementL = assertionL . assContents . sasStatements . _nlhead
+
+userRefL :: Getter AuthnResponse UserRef
+userRefL = to $ \aresp ->
```

Comment:
If it would be more convenient for you to have an actual `Lens'` here, you can get that with http://hackage.haskell.org/package/lens-4.17/docs/Control-Lens-Unsound.html#v:lensProduct

Reply:
I'm even 60% sure it will work!

Reply:
it would not.

i was wondering if you would suggest a solution here, though!  :)

Reply:
Oh well